src/main/java/org/opensearch/knn/index/fetch/KNNFetchSubPhase.java

```diff
@@ -0,0 +1,226 @@
+            XContentBuilder builder = new XContentBuilder(XContentType.JSON.xContent(), streamOutput);
+            builder.value(maps);
+            hitContext.hit().sourceRef(BytesReference.bytes(builder));
+        }
+
+        protected void syntheticNestedDocValues(MapperService mapperService, HitContext hitContext, Map<String, Object> maps)
+            throws IOException {
+            DocumentMapper documentMapper = mapperService.documentMapper();
+            Map<String, ObjectMapper> mapperMap = documentMapper.objectMappers();
+            SearchHit hit = hitContext.hit();
+
+            for (ObjectMapper objectMapper : mapperMap.values()) {
+                if (objectMapper == null) {
+                    continue;
+                }
+                if (!objectMapper.nested().isNested()) {
+                    continue;
+                }
+                String path = objectMapper.fullPath();
+                for (DocValueField f : fields) {
+                    if (!f.field.startsWith(path)) {
+                        continue;
+                    }
+                    if (!maps.containsKey(path)) {
+                        continue;
+                    }
+
+                    // path to nested field:
+                    Object nestedObj = maps.get(path);
+                    if (!(nestedObj instanceof ArrayList)) {
+                        continue;
+                    }
+                    // nested array in one nested path
+                    ArrayList nestedDocList = (ArrayList) nestedObj;
+
+                    log.debug(
+                        "object mapper: nested:"
+                            + objectMapper.nested().isNested()
+                            + " Value:"
+                            + objectMapper.fullPath()
+                            + " field:"
+                            + f.field
+                    );
+
+                    Query parentFilter = Queries.newNonNestedFilter();
+                    Query childFilter = objectMapper.nestedTypeFilter();
+                    ContextIndexSearcher searcher = fetchContext.searcher();
+                    final Weight childWeight = searcher.createWeight(searcher.rewrite(childFilter), ScoreMode.COMPLETE_NO_SCORES, 1f);
+
+                    LeafReaderContext subReaderContext = hitContext.readerContext();
+                    Scorer childScorer = childWeight.scorer(subReaderContext);
+                    DocIdSetIterator childIter = childScorer.iterator();
+                    BitSet parentBits = fetchContext.getQueryShardContext().bitsetFilter(parentFilter).getBitSet(subReaderContext);
+
+                    int currentParent = hit.docId() - subReaderContext.docBase;
+                    int previousParent = parentBits.prevSetBit(currentParent - 1);
+                    int childDocId = childIter.advance(previousParent + 1);
+                    SourceLookup nestedVecSourceLookup = new SourceLookup();
```

Comment:
We are doing the query again with this piece of code. Can we somehow avoid this? Because it will add latency in overall fetch phase. 

One thing I can think of is, the way docIds are created for Parent and chid is: If a parent document has 3 child fields, then child docIds will be 1, 2, 3 and parent docId will be 4. So we can have access to how many children a parent docId has we can easily avoid this extra query.

Reply:
This code reference from `FetshPhase#getInternalNestedIdentity`. i think it for filtering docid which is parent and child. we need find every parent doc and get the child doc which id between in the `parentDoc` and `parentDoc -1` . 

and also i am not sure when there is `soft_deletes` enabled, we only assumed doc_ids as sequentially with out score check, whether or not get the deleted doc